### ISSUE TITLE
Distances on GT boundaries

### ISSUE THREAD

**Issue by maxmanko** [LIVIAETS/boundary-loss#8]
Thank you for your loss function, it exceeded all my expectations. I applied it to solving the problem of pulmonary nodules segmentation, implementing this loss function in Keras framework.

You reset the distances for pixels that are on the edges of GT:
`res [c] = distance (negmask) * negmask - (distance (posmask) - 1) * posmask`
Thus, these pixels are not included in the loss function. Would it be better to take these pixels into account in the loss function?
`res [c] = distance (negmask) * negmask - distance (posmask) * posmask`

**Comment by HKervadec** [comment]
The rationale, when I implemented it, was that we compute the distance _to the boundary_. By this logic, the distance of the boundary wrt the boundary should be 0. Since this not the behavior of the SciPy function [distance_transform_edt](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.distance_transform_edt.html). I did that manually by shifting the distance by 1 on the distance to the positive mask. 

To be fair, I do not think it would change (much) the results, since it is only shifting the distance values inside the object. But I did not had time to compare the impact of this implementation detail.

**Comment by JunMa11** [comment]
Hi @maxmanko , would it be possible for you to share your keras implementation of surface loss?

**Comment by zye1996** [comment]
> The rationale, when I implemented it, was that we compute the distance _to the boundary_. By this logic, the distance of the boundary wrt the boundary should be 0. Since this not the behavior of the SciPy function [distance_transform_edt](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.distance_transform_edt.html). I did that manually by shifting the distance by 1 on the distance to the positive mask.
> 
> To be fair, I do not think it would change (much) the results, since it is only shifting the distance values inside the object. But I did not had time to compare the impact of this implementation detail.

Hi, I have a question about the -1 shift here. If the distance of boundary wrt the boundary is 0, then the boundary would not contribute to loss and thus is not supervised. Will it affect the performance of small objects?

**Comment by HKervadec** [comment]
Hello,

> Will it affect the performance of small objects?

Now that you mention it, it could be a possiblity on very small objects (such as very thin blood vessels). In that case, it could be worth your time to compare the two; with and without that shift. But I think as long as you have an area with a negative distance map, you should be good.

(That could actually be a new assertion to be included into the function ; that some negative area exists.)

Best,

Hoel
